### PR TITLE
refactor async DoFns

### PR DESCRIPTION
**Pull request by nevillelyh**

### CODE REVIEW

File: scio-core/src/main/java/com/spotify/scio/transforms/BaseAsyncDoFn.java [hunk 1]
```diff
@@ -64,9 +61,10 @@ public void finishBundle(FinishBundleContext c) {
         waitForFutures(futures.values());
       } catch (InterruptedException e) {
         Thread.currentThread().interrupt();
+        LOG.error("Failed to process futures", e);
         throw new RuntimeException("Failed to process futures", e);
       } catch (ExecutionException e) {
-        throw new RuntimeException("Failed to process futures", e);
+        LOG.error("Failed to process futures", e);
```
Comment (ClaireMcGinty) [inline scio-core/src/main/java/com/spotify/scio/transforms/BaseAsyncDoFn.java]: Why take away the RuntimeException? Shouldn't the bundle get failed ?

File: scio-core/src/main/java/com/spotify/scio/transforms/BaseAsyncLookupDoFn.java [hunk 2]
```diff
@@ -250,14 +255,15 @@ public boolean isSuccess() {
   }
 
   /**
-   * Supplier for {@link AsyncLookupDoFn} with caching logic.
+   * Supplier for {@link BaseAsyncLookupDoFn} with caching logic.
```
Comment (ClaireMcGinty) [inline scio-core/src/main/java/com/spotify/scio/transforms/BaseAsyncLookupDoFn.java]: Isn't this technically supplying a `Cache` not a `BaseAsyncLookupDoFn` ? 🤔 Also -- it seems a bit strange to me that this class is both a Supplier for new `Cache`s while also handling get/put logic for a cache of `Cache`s? It seems like `CacheManager` might be a more accurate name... it's not a big deal though.

File: scio-core/src/main/java/com/spotify/scio/transforms/FutureHandlers.java [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+import java.util.concurrent.ExecutionException;
+import java.util.function.Function;
+
+/** Utility to abstract away Guava, Java 8 and Scala future handling. */
+public class FutureHandlers {
+
+  /**
+   * Base interface for future handling.
+   * @param <F> future type.
+   * @param <T> value type.
```
Comment (ClaireMcGinty) [inline scio-core/src/main/java/com/spotify/scio/transforms/FutureHandlers.java]: jw, why is this parameterized as `<T>` ? It seems like `<B>` would be more consistent with the rest of the package unless I'm misreading this!

File: scio-core/src/main/java/com/spotify/scio/transforms/BaseAsyncDoFn.java (hunk 1 above)
Comment (nevillelyh) [reply scio-core/src/main/java/com/spotify/scio/transforms/BaseAsyncDoFn.java]: You're right.
I had to remove it for the `BaseAsyncLookupDoFn` case since we don't want to fail a bundle if there're lookup errors. In that case the callback appends into the error queue and we surface it to user as a failed `Try`.
In this case we should rethrow.

File: scio-core/src/main/java/com/spotify/scio/transforms/FutureHandlers.java (hunk 3 above)
Comment (nevillelyh) [reply scio-core/src/main/java/com/spotify/scio/transforms/FutureHandlers.java]: Agree, was just being lazy :sob:
Although `B` is usually for cases like `<A, B, C>`. Call it `V` instead?

Comment (ClaireMcGinty) [reply scio-core/src/main/java/com/spotify/scio/transforms/FutureHandlers.java]: `V` sounds good! 👍

File: scio-core/src/main/java/com/spotify/scio/transforms/BaseAsyncLookupDoFn.java (hunk 2 above)
Comment (nevillelyh) [reply scio-core/src/main/java/com/spotify/scio/transforms/BaseAsyncLookupDoFn.java]: Maybe rephrase it as `{@link Cache} supplier for {@link BaseAsyncLookupDoFn}`? Supplier is from the Java naming convention for `() => T`.

Comment (ClaireMcGinty) [reply scio-core/src/main/java/com/spotify/scio/transforms/BaseAsyncLookupDoFn.java]: oh, yeah my thought here was more that the scope of this class seems to be greater than just being a `Supplier`! I think your rephrasing makes sense though.
